core/state-machine/src/basic.rs

```diff
@@ -18,67 +18,64 @@
 
 use std::collections::HashMap;
 use std::iter::FromIterator;
+use crate::backend::{Backend, InMemory};
 use hash_db::Hasher;
 use trie::trie_root;
 use primitives::offchain;
-use primitives::storage::well_known_keys::{CHANGES_TRIE_CONFIG, CODE, HEAP_PAGES};
+use primitives::storage::well_known_keys::{HEAP_PAGES, is_child_storage_key};
 use parity_codec::Encode;
-use super::{ChildStorageKey, Externalities, OverlayedChanges};
+use super::{ChildStorageKey, Externalities};
 use log::warn;
 
 /// Simple HashMap-based Externalities impl.
+#[derive(Debug)]
 pub struct BasicExternalities {
-	inner: HashMap<Vec<u8>, Vec<u8>>,
-	changes: OverlayedChanges,
-	code: Option<Vec<u8>>,
+	top: HashMap<Vec<u8>, Vec<u8>>,
+	children: HashMap<Vec<u8>, HashMap<Vec<u8>, Vec<u8>>>,
 }
 
 impl BasicExternalities {
 	/// Create a new instance of `BasicExternalities`
-	pub fn new(inner: HashMap<Vec<u8>, Vec<u8>>) -> Self {
-		Self::new_with_code(&[], inner)
+	pub fn new(top: HashMap<Vec<u8>, Vec<u8>>) -> Self {
+		Self::new_with_children(top, Default::default())
 	}
 
-	/// Create a new instance of `BasicExternalities`
-	pub fn new_with_code(code: &[u8], mut inner: HashMap<Vec<u8>, Vec<u8>>) -> Self {
-		let mut overlay = OverlayedChanges::default();
-		super::set_changes_trie_config(
-			&mut overlay,
-			inner.get(&CHANGES_TRIE_CONFIG.to_vec()).cloned(),
-			false,
-		).expect("changes trie configuration is correct in test env; qed");
-
-		inner.insert(HEAP_PAGES.to_vec(), 8u64.encode());
-
+	/// Create a new instance of `BasicExternalities` with children
+	pub fn new_with_children(
+		mut top: HashMap<Vec<u8>, Vec<u8>>,
+		children: HashMap<Vec<u8>, HashMap<Vec<u8>, Vec<u8>>>,
+	) -> Self {
+		top.insert(HEAP_PAGES.to_vec(), 8u64.encode());
 		BasicExternalities {
-			inner,
-			changes: overlay,
-			code: Some(code.to_vec()),
+			top,
+			children,
 		}
 	}
 
 	/// Insert key/value
 	pub fn insert(&mut self, k: Vec<u8>, v: Vec<u8>) -> Option<Vec<u8>> {
-		self.inner.insert(k, v)
+		self.top.insert(k, v)
 	}
-}
 
-impl ::std::fmt::Debug for BasicExternalities {
-	fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
-		write!(f, "{:?}", self.inner)
+	/// Get storages
+	pub fn storages(self) -> (
```

Comment:
```suggestion
	pub fn into_storages(self) -> (
```

Or maybe `into_inner`.